pymeasure/instruments/hp/hp8116a.py

```diff
@@ -0,0 +1,176 @@
+        'pulse': 'W4',
+    }
+
+    @property
+    def operating_mode(self):
+        """ The operating mode of the instrument.
+        """
+        raise NotImplementedError
+    
+    @operating_mode.setter
+    def operating_mode(self, mode):
+        mode_set = self.OPERATING_MODES[strict_discrete_set(mode, self.OPERATING_MODES)]
+        self.write(mode_set)
+
+    @property
+    def control_mode(self):
+        """ The control mode of the instrument.
+        """
+        raise NotImplementedError
+    
+    @control_mode.setter
+    def control_mode(self, mode):
+        mode_set = self.CONTROL_MODES[strict_discrete_set(mode, self.CONTROL_MODES)]
+        self.write(mode_set)
+    
+    @property
+    def trigger_slope(self):
+        """ The trigger slope of the instrument.
+        """
+        raise NotImplementedError
+    
+    @trigger_slope.setter
+    def trigger_slope(self, slope):
+        slope_set = self.TRIGGER_SLOPE[strict_discrete_set(slope, self.TRIGGER_SLOPE)]
+        self.write(slope_set)
+    
+    @property
+    def shape(self):
+        """ The shape of the waveform.
+        """
+        raise NotImplementedError
+    
+    @shape.setter
+    def shape(self, shape):
+        shape_set = self.SHAPES[strict_discrete_set(shape, self.SHAPES)]
+        self.write(shape_set)
+    
+    @property
+    def haversine(self):
+        """ With triggered, gate or E.BUR operating mode selected and shape
+        set to sine or triangle, this setting shifts the start phase of the
+        waveform is -90°. As a result, haversine and havertriangle signals
+        can be generated.
+        """
+        raise NotImplementedError
+    
+    @haversine.setter
+    def haversine(self, haversine):
+        haversine_set = int(haversine)
+        haversine_cmd = "Z" + str(int(strict_discrete_set(haversine_set, [0, 1])))
+        self.write(haversine_cmd)
+
+    @property
+    def frequency(self):
+        """ A floating point property that controls the frequency of the
+        output in Hz. The allowed frequency range is 1 mHz to 52.5 MHz.
+        The resolution is 3 digits.
+        """
+        raise NotImplementedError
+    
+    @frequency.setter
+    def frequency(self, frequency):
+        frequency = float(strict_range(frequency, [1e-3, 52.5e6]))
+
+        # Send frequency in mHz, Hz, kHz or MHz as appropriate
+        if frequency < 1:
+            freq_cmd = f'FRQ {frequency*1e3:.3g} MZ'  # mHz
+        elif frequency < 1e3:
+            freq_cmd = f'FRQ {frequency:.3g} HZ'
+        elif frequency < 1e6:
+            freq_cmd = f'FRQ {frequency*1e-3:.3g} KHZ'
+        else:
+            freq_cmd = f'FRQ {frequency*1e-6:.3g} MHZ'
+        
+        self.write(freq_cmd)
+
+    @property
+    def amplitude(self):
+        """ A floating point property that controls the amplitude of the
+        output in V. The allowed amplitude range is -10 V to 10 V.
+        """
+        raise NotImplementedError
+
+
```

Comment:
@CodingMarco, thank for your draft contribution.
I understand you still need to complete it, but as an early comment, I was curious whether there is a specific reason to avoid usage of property factories like `Instrument.control` etc.

Reply:
The problem is that this is another instrument from the pre-SCPI era. Not only that, things are even more difficult since some units of this instrument have the GPIB "EOI" line cut which results in a timeout on all reads with pyvisa (see service note 8116A-07A "Solving EOI Prablems" in the 8116A service manual). Because of that, only `adapter.read_bytes(nr_of_bytes)` can be used to read the instrument state or query specific parameters (nr_of_bytes is different for each command).
Also it is important to have a short delay between writing a command and reading the response, otherwise the 8116A answert "NO MESSAGE".

Would this be possible with `Instrument.control`?

Reply:
I think it's possible, all the problems you mentioned are related to the physical access to the instrument while `Instrument.control` is in charge of formatting the messages to instrument and decoding the  messages from instrument in a flexible and standardized way.

Regarding your specific points:
- non SCPI: This possible, have a look at the work don't by @LongnoseRob in pymeasure/instruments/hp/hp3478A.py
- For the EOI and delay problem, it should be possible to customize `write`, `read` and `ask` methods of `Instrument` class to serve your needs

Hope this helps

Reply:
Thanks for the suggestion, as you can see I have made extensive use of the predefined properties which were very helpful in the end. However, I had to essentially copy `values()` from `adapter` to my class because the method in adapter calls `self.ask()` which I have to override. Maybe this is a general design issue that can be addressed in the future.